futures-util/src/future/pinned.rs

```diff
@@ -0,0 +1,116 @@
+//! Definition of the `PinnedFut` adapter combinator
+
+// TODO: import `Pinned` and use it to make `Borrowed` immovable.
+use core::mem::PinMut;
+
+use futures_core::{Future, Poll};
+use futures_core::task;
+
+pub trait PinnedFnLt<'a, Data: 'a, Output> {
+    type Future: Future<Output = Output> + 'a;
+    fn apply(self, data: PinMut<'a, Data>) -> Self::Future;
+}
+
+pub trait PinnedFn<Data, Output>: for<'a> PinnedFnLt<'a, Data, Output> {}
+impl<Data, Output, T> PinnedFn<Data, Output> for T
+    where T: for<'a> PinnedFnLt<'a, Data, Output> {}
+
+impl<'a, Data, Output, Fut, T> PinnedFnLt<'a, Data, Output> for T
+where
+    Data: 'a,
+    T: FnOnce(PinMut<'a, Data>) -> Fut,
+    Fut: Future<Output = Output> + 'a,
+{
+    type Future = Fut;
+    fn apply(self, data: PinMut<'a, Data>) -> Self::Future {
+        (self)(data)
+    }
+}
+
+/// A future which borrows a value for an asynchronous lifetime.
+///
+/// Created by the `borrowed` function.
+#[must_use = "futures do nothing unless polled"]
+#[allow(missing_debug_implementations)]
+pub struct PinnedFut<'any, Data: 'any, Output, F: PinnedFn<Data, Output> + 'any> {
```

Comment:
I can't actually make this work because I get stuck on that `PinnedFut` isn't allowed to outlive `'any` no matter what I do. This seems like I need `#[may_dangle]`, but somehow stronger. I don't even need a lifetime to be part of the type-- I just need a lifetime that does not outlive `Data` (can't always use `'static`) but which will unify with other lifetimes.

Reply:
Even using 'static doesn't work-- then the WF checks for HRTB aren't kicking in for some reason: i get errors that `T: for<'a> PinnedFnLt<'a, Data, Output>` isn't satisfied if I add a `where Self: 'a` bound to `PinnedFnLt`.